### PR TITLE
fix(sql): clear all fields of execution plan objects to avoid unexpected errors

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Lack of proper field reset was leading to unexpected exceptions thrown in `SqlCodeGenerator`.

Also adds a test to make `QueryModel#clear()` less fragile.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 172 (57.56%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryColumn.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|1|10|10.00%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceIntHashMap.java|8|20|40.00%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceObjHashMap.java|8|20|40.00%|
|:large_blue_circle:|io/questdb/std/IntHashSet.java|6|11|54.55%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceHashSet.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|69|96|71.88%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 171 (57.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryColumn.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|1|10|10.00%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceIntHashMap.java|8|20|40.00%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceObjHashMap.java|8|20|40.00%|
|:large_blue_circle:|io/questdb/std/IntHashSet.java|6|11|54.55%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceHashSet.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|69|95|72.63%|

**Comment by bziobrowski** [comment]
Looks fine to me but it'd be good to increase coverage .

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java [hunk 1]
```diff
@@ -306,7 +315,10 @@ public void clear() {
         clearSampleBy();
         orderBy.clear();
         orderByDirection.clear();
+        orderByAdvice.clear();
+        orderByDirectionAdvice.clear();
         orderByPosition = 0;
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/model/QueryModel.java:320]: Have you considered using OrderByMnemonic.ORDER_BY_UNKNOWN ?

**Comment by puzpuzpuz** [comment]
> Looks fine to me but it'd be good to increase coverage .

Increase coverage of which class? `QueryModel` has zero unit tests before this PR and this patch adds a single test aimed to verify that `clear()` resets all fields. So, this PR already increases the coverage.

**Comment by bziobrowski** [comment]
I meant other classes changed in the PR . The coverage  is below 60% so it's a bit lower than usual . 
I know that these equals & hashCode methods were added to support the check but since they're here someone might use them in the future .

**Comment by puzpuzpuz** [comment]
> The coverage is below 60% so it's a bit lower than usual

The reported coverage is incorrect. Try running corresponding unit tests in IDEA.

I've added a basic test for equals and hashCode implementations in 3e4c8c6371809ef82b91d1717add1ebba17c14b9, as well as included some fixes of the hashCode methods.

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/model/QueryModel.java:320]: `orderByPosition` here stands for the position in the query text. Did you mean to use `OrderByMnemonic.ORDER_BY_UNKNOWN` in another line or I'm missing anything?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 135 / 181 (74.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryColumn.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|1|10|10.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|69|95|72.63%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceIntHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceObjHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceHashSet.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/std/IntHashSet.java|14|15|93.33%|

File: core/src/main/java/io/questdb/griffin/model/QueryModel.java (hunk 1 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/model/QueryModel.java:320]: I think code changed and line is not correct anymore . I meant 
```
orderByAdviceMnemonic = 0;
```

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/model/QueryModel.java:321]: My bad. Addressed in 5ad7205df202437b7c66622de11200ee569a4777

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 136 / 182 (74.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryColumn.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|1|10|10.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|70|96|72.92%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceIntHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceObjHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceHashSet.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/std/IntHashSet.java|14|15|93.33%|

**Comment by puzpuzpuz** [comment]
> Any idea why coverage might be incorrect ?

It has something to do with the plugin we use. Previous coverage report had something like 40-50% coverage reported for `LowerCaseCharSequenceIntHashMap` and `LowerCaseCharSequenceObjHashMap` while the actual values were 70+%.

**Review by bziobrowski — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
@bziobrowski thanks for the code review!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 136 / 182 (74.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryColumn.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|1|10|10.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|70|96|72.92%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceIntHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceObjHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceHashSet.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/std/IntHashSet.java|14|15|93.33%|

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
give me time to review, I would like to comprehend why we need such monstrous hash code functions

**Comment by puzpuzpuz** [comment]
> give me time to review, I would like to comprehend why we need such monstrous hash code functions

No worries, take the time you need. `hashCode` methods come along with `equals` and the latter is necessary to have tests around `QueryModel#clear()` which is quite error-prone.

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 136 / 182 (74.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryColumn.java|0|5|00.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|1|10|10.00%|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|70|96|72.92%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceIntHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceObjHashMap.java|19|21|90.48%|
|:large_blue_circle:|io/questdb/std/LowerCaseCharSequenceHashSet.java|13|14|92.86%|
|:large_blue_circle:|io/questdb/std/IntHashSet.java|14|15|93.33%|

**Pull request merged**